scale: revert back to one single join switch

cathy-zhou opened the pull request:
We used to have single join switch and commit c372 changed to use
per-node join switch to reduce logical flows in the SB's
lr_in_arp_resolve table.

Since then, there are several OVN patches to provide options to change
the default behavior of OVN and reduce the logical flows in the SB.
In particular, the "dynamic_neigh_routers" option can be set to "true"
and the "learn_from_arp_request" option can be set to "false" to avoid
static neighbor flows and diabling neighbor learning from ARP request,
which minimizes the SB logical flows explosion.

With all the patches, we can now revert back to single join switch
implementation. Furture changes will be needed to set those options to
the gateway routers once OVN patches are avaible.

Signed-off-by: Yun Zhou <yunz@nvidia.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/ovn-org/ovn-kubernetes/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Trivial changes are exempt from following this template.
If your change is non-trivial, please provide the following information:
-->

**- What this PR does and why is it needed**
<!--
A summary of the changes within this pull request and some context
as to why they were made
-->

**- Special notes for reviewers**
<!--
What exactly did you change - you may also defer to information
contained in commit messages. At a bare minimum it's worth highlighting
which areas of the code were changed as it's easier to assign reviewers
-->


**- How to verify it**
<!--
Did you include unit tests? or end-to-end tests?
How can I manually verify that this patch achieves its objective
-->


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

trozet commented:
/hold we still don't even have the relevant OVN fixes upstream in our CI builds

girishmg commented:
@cathy-zhou can you please rebase and include the required changes for EgressIP and EgressFirewall fixes?

cathy-zhou commented:
@girishmg thanks for your comments. I've updated my code and repushed.

girishmg commented:
@dcbw @danwinship @trozet can we get some eyes on this please? We have been running this downstream for close to two months now, and are happy with how it is performing.

danwinship commented:
> With all the patches, we can now revert back to single join switch
> implementation. Furture changes will be needed to set those options to
> the gateway routers once OVN patches are avaible.

If the code is going to depend on those options being set, shouldn't we set them at startup?

danwinship commented on go-controller/pkg/util/net.go:
```diff
@@ -100,6 +100,31 @@ func GetPortAddresses(portName string, ovnNBClient goovn.Client) (net.HardwareAd
+			"stderr: %q, error: %v", portName, stderr, err)
+	}
+
+	// eg: `["100.64.1.1/29", "fd98:1::/125"]`
```
// eg: `["100.64.0.20/16", "fd98::14/64"]`

cathy-zhou commented on go-controller/pkg/ovn/egressgw.go:
```diff
@@ -339,16 +339,15 @@ func (oc *Controller) addHybridRoutePolicyForPod(podIP net.IP, node string) erro
 		// get the GR to join switch ip address
-		out, stderr, err := util.RunOVNNbctl("--data=bare", "--no-heading", "--columns=networks", "find",
-			"logical_router_port", fmt.Sprintf("name=rtoj-GR_%s", node))
+		grJoinIPnets, err := util.GetLrpNetworks(fmt.Sprintf("rtoj-GR_%s", node))
```
will fix.

cathy-zhou commented on go-controller/pkg/ovn/logical_switch_manager.go:
```diff
@@ -261,3 +283,131 @@ func (manager *logicalSwitchManager) ReleaseIPs(nodeName string, ipnets []*net.I
+		lrpIPCacheLock: sync.Mutex{},
+	}
+	var joinSubnets []*net.IPNet
+	for _, joinSubnetString := range []string{v4JoinSubnetCidr, v6JoinSubnetCidr} {
```
yes, we intentionally did that.

cathy-zhou commented on go-controller/pkg/ovn/logical_switch_manager.go:
```diff
@@ -261,3 +283,131 @@ func (manager *logicalSwitchManager) ReleaseIPs(nodeName string, ipnets []*net.I
+}
+
+// ensureJoinLRPIPs tries to allocate the LRP IPs if it is not yet allocated, then they will be stored in the cache
+func (jsIPManager *joinSwitchIPManager) ensureJoinLRPIPs(nodeName string) (gwLRPIPs []*net.IPNet, err error) {
```
I am not sure I understand your concern. Do you see a possible race? Can you elaborate?

danwinship replied:
I don't see any particular race, but the code doesn't make it obvious that there's _not_ a race, whereas if there was only a single lock, that would be clearer

cathy-zhou replied:
@danwinship Thanks. I looked it over again, and I still don't see any possible race here. Since joinSwitchIPManager.lrpIPCacheLock, logicalSwitchManager and ipam all has a lock (note logicalSwitchManager's lock is not useful here but we'd just like to reuse the existing code), I am reluctant to introduce another level of lock here unless it is necesarry.

trozet replied:
It looks like this is a side effect of inheriting the logical switch manager with its own locks, which then calls an ipam object which also has a lock. It looks like the code should be solid as is. I agree the more locks, the more potential pitfalls, but I don't see a good way around it with the code as is. We can perhaps refactor it in the future if it becomes a bigger problem.

cathy-zhou commented on go-controller/pkg/ovn/master.go:
```diff
@@ -323,60 +309,55 @@ func (oc *Controller) SetupMaster(masterNodeName string) error {
-		return fmt.Errorf("failed to marshal node %q join subnets annotation for subnet %s",
-			node.Name, util.JoinIPNets(subnets, ","))
-	}
-	err = oc.kube.SetAnnotationsOnNode(node, nodeAnnotations)
```
No, this change does not have code to enabling upgrade the existing per-node join switch DB to single join switch topology. Will add that support.

cathy-zhou commented on go-controller/pkg/ovn/master.go:
```diff
@@ -323,60 +309,55 @@ func (oc *Controller) SetupMaster(masterNodeName string) error {
+	// Connect the distributed router to ovnJoinSwitch.
+	drSwitchPort := joinSwitchToGwRouterPrefix + ovnClusterRouter
+	drRouterPort := gwRouterToJoinSwitchPrefix + ovnClusterRouter
+	drLRPMAC := util.IPAddrToHWAddr(gwLRPIPs[0].IP)
```
fixed. Yes, unfortunately we compute the mac whenever needed as I don't like the idea of taking additional gwLRPMAC arguments into gatewayInit() function. Do you have any suggestions?

cathy-zhou commented on go-controller/pkg/ovn/master.go:
```diff
@@ -440,18 +421,18 @@ func (oc *Controller) syncNodeManagementPort(node *kapi.Node, hostSubnets []*net
 	}
 
-	err = gatewayInit(node.Name, clusterSubnets, hostSubnets, joinSubnets, l3GatewayConfig, oc.SCTPSupport)
+	drLRPIPs, _ := oc.joinSwIPManager.getJoinLRPCacheIPs(ovnClusterRouter)
```
We intentionally did this. we used to have the code to make the joinSwIPManager reserve the .1 IP for ovnClusterRouter but we think it is cleaner to just call getJoinLRPCacheIPs() even for drLRPIPs.

cathy-zhou commented on go-controller/pkg/ovn/master.go:
```diff
@@ -951,7 +965,6 @@ func (oc *Controller) syncNodes(nodes []interface{}) {
 	type NodeSubnets struct {
 		hostSubnets []*net.IPNet
-		joinSubnets []*net.IPNet
 	}
```
sure

cathy-zhou commented:
@danwinship Other than several comments I don't fully understand, I've addressed all of your comments and repushed. Can you please take a look. Thanks

trozet commented on go-controller/pkg/ovn/master.go:
```diff
@@ -910,6 +891,37 @@ func (oc *Controller) syncNodesPeriodic() {
 	deleteChassis(oc.ovnSBClient, chassisMap)
 }
 
+func (oc *Controller) getJoinLRPAddresses(nodeName string) []*net.IPNet {
```
should this function really be in this file?

cathy-zhou replied:
can you please suggest where it should go? It requires "oc" so it cannot be an utility function.

trozet replied:
maybe gateway.go? That's where the joinSwitch prefixes are. Also add a line of doc text explaining what the function does.

cathy-zhou replied:
sure. fixed

trozet commented on go-controller/pkg/util/net.go:
```diff
@@ -100,6 +74,31 @@ func GetPortAddresses(portName string, ovnNBClient goovn.Client) (net.HardwareAd
+	// eg: `["100.64.0.3/16", "fd98::3/64"]`
+	output = strings.Trim(output, "[]")
+	if output != "" {
+		for _, ipNetStr := range strings.Split(output, ", ") {
```
ok for now, but we need to move this kind of stuff to go bindings

trozet commented on go-controller/pkg/ovn/master.go:
```diff
@@ -155,6 +233,12 @@ func (oc *Controller) StartClusterMaster(masterNodeName string) error {
 		return err
 	}
 
+	err = oc.upgradeOVNTopology(existingNodes)
```
If you push another patch, do you mind fixing like 232?

klog.Errorf("Error in initializing/fetching subnets: %v", err)

should be "Error in fetching nodes"

trozet commented on go-controller/pkg/ovn/master.go:
```diff
@@ -121,6 +124,81 @@ func (oc *Controller) Start(kClient kubernetes.Interface, nodeName string, wg *s
+		if strings.Contains(stderr, "no row") {
+			// no ovnClusterRouter exists, DB is empty, nothing to upgrade
+			ver = OvnCurrentTopologyVersion
+		}
```
what about an else here? log a warning? "unable to detect k8s-ovn-topo-version, will assume upgrade is needed"

cathy-zhou replied:
sure. I will log an info instead of warning though, since it is expected as this is the first time we introduce k8s-ovn-topo-version.

trozet replied:
can you also add that logging message here please in the next patch set?

cathy-zhou replied:
sorry that I missed this. Added now.

trozet commented on go-controller/pkg/ovn/master.go:
```diff
@@ -121,6 +124,81 @@ func (oc *Controller) Start(kClient kubernetes.Interface, nodeName string, wg *s
+	}
+
+	nodeSwitches, stderr, err := util.RunOVNNbctl("--data=bare", "--no-heading",
+		"--columns=name,other-config", "find", "logical_switch", "other-config:subnet!=_")
```
what is the significance of matching on "other-config:subnet!=_" ?

cathy-zhou replied:
Because there are logical switches (join, ext-<nodeName etc.) that do not have other-config options, the output of the command would not be as expected and parsing will be incorrect. I was trying to filter out those swithes. But my changes are not correct either, as the node switch may not have "other-config:ipv6_prefix" set, not "other-config:subnet".

trozet commented on go-controller/pkg/ovn/gateway_cleanup.go:
```diff
@@ -159,3 +159,96 @@ func staticRouteCleanup(nextHops []net.IP) {
+			"stderr: %q, error: %v", nodeName, stderr, err)
+	}
+
+	// If exists, remove the TCP, UDP load-balancers created for north-south traffic for gateway router.
```
Is this really required? This is configured on a per GR basis, and we have some sync logic when services come up to look for stale reject ACLs. I think for that we need the load balancers to exist.

cathy-zhou replied:
The upgradeToSingleSwitchOVNTopology() function is used to do two things:

trozet replied:
This logic here: https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/service.go#L104

I don't see why for upgrade we need to delete anything other than the join switch and the join switch ports. Those load balancers are also configured on the node logical switches. I think we need to only delete what is required.

cathy-zhou replied:
Ok, I added an argument to the multiJoinSwitchGatewayCleanup() function to differentiate the upgrade case and syncNode() equivalence case. Please take a look and see if it is ok. Thanks

trozet replied:
Thanks

trozet commented on go-controller/pkg/ovn/gateway_cleanup.go:
```diff
@@ -159,3 +159,96 @@ func staticRouteCleanup(nextHops []net.IP) {
+			"stderr: %q, error: %v", "join_"+nodeName, stderr, err)
+	}
+
+	// Remove the gateway router associated with nodeName
```
why do we need to delete the logical router? or the external switch? Only thing that needs to be disconnected is the join switch, then the link between join switch and the GR reconfigured.

cathy-zhou replied:
same as above. Thanks

trozet commented:
@cathy-zhou can you comment on what version of OVN this was introduced in, not just the commit, but the RPM version?

girishmg commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -287,6 +259,19 @@ func gatewayInit(nodeName string, clusterIPSubnet []*net.IPNet, hostSubnets []*n
+	// cluster router, to ensure that the return traffic goes back
+	// to the same gateway router
+	for _, gwLRPIP := range gwLRPIPs {
+		stdout, stderr, err = util.RunOVNNbctl("--may-exist", "lr-route-add",
```
@trozet reply packets from the pod to external requests should be send back to the GR from where the request packets ingress the OVN logical topology.  If you don't have this /32 routes, then the reply packet from a Pod will be sent to GR on the pod's node. To achieve symmetric routing this is required. 

We didn't need it with multiple join switches because we use /29 subnet for each of the join switch. Due to longest prefix match rule, the /29 routes in ovn_cluster_router will get matched before the source-ip based routes.

trozet replied:
The incoming packet would be SNAT'ed to the GR IP address. When the reply packet comes back, it is already destined for to the right GR. ovn_cluster_router is directly attached to the subnet where that GR resides. It doesn't need a static route, it already knows where to send the packet because it has a directly connected route to that subnet. It just arps and sends the packet directly to the GR. If this really is required, it seems like a bug in OVN to me. Ping me on slack and we can discuss further.

trozet replied:
After talking it through with Girish on Friday he clarified why we need this. I went ahead and filed: https://bugzilla.redhat.com/show_bug.cgi?id=1891516

@cathy-zhou can you please add a comment here to remove this route after 1891516 is fixed?

cathy-zhou replied:
done. thanks @trozet

cathy-zhou commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -123,8 +81,22 @@ func gatewayInit(nodeName string, clusterIPSubnet []*net.IPNet, hostSubnets []*n
 		}
 	}
 
+	stdout, stderr, err = util.RunOVNNbctl("set", "logical_router", gatewayRouter,
```
sure

cathy-zhou commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -123,8 +81,22 @@ func gatewayInit(nodeName string, clusterIPSubnet []*net.IPNet, hostSubnets []*n
+			"stdout: %q, stderr: %q, error: %v", gatewayRouter, stdout, stderr, err)
+	}
+
+	stdout, stderr, err = util.RunOVNNbctl("set", "logical_router",
```
sure

cathy-zhou commented on go-controller/pkg/ovn/logical_switch_manager.go:
```diff
@@ -167,13 +167,35 @@ func (manager *logicalSwitchManager) AllocateIPs(nodeName string, ipnets []*net.
+
+	var err error
+	allocated := make(map[int]*net.IPNet)
+	defer func() {
```
ok, I will move this change to a seperate commit.

cathy-zhou commented:
I beleive the two OVN options are in v20.06.2 and 20.09.0

trozet commented on go-controller/pkg/ovn/gateway_cleanup.go:
```diff
@@ -159,3 +159,109 @@ func staticRouteCleanup(nextHops []net.IP) {
+	}
+
+	// Remove the patch port on the distributed router that connects to join switch
+	_, stderr, err = util.RunOVNNbctl("--if-exist", "lrp-del", "dtoj-"+nodeName)
```
Need to also delete the lrp on the GR

cathy-zhou replied:
thanks for catching this. Fixed

trozet commented on go-controller/pkg/ovn/master.go:
```diff
@@ -121,6 +124,83 @@ func (oc *Controller) Start(kClient kubernetes.Interface, nodeName string, wg *s
+		}
+	}
+
+	// If current DB version is greater than OvnSingleJoinSwithTopoVersion, no need to upgrade to single switch topoloby
```
topology*

cathy-zhou replied:
fixed

trozet commented on go-controller/pkg/ovn/master.go:
```diff
@@ -121,6 +124,83 @@ func (oc *Controller) Start(kClient kubernetes.Interface, nodeName string, wg *s
 }
 
+// delete obsoleted logical OVN entities that are specific for Multiple join switches OVN topology. Also cleanup
+// OVN entities for deleted nodes (similiar to syncNodes() but for obsoleted Multiple join switches OVN topology
```
similar*

cathy-zhou replied:
fixed

trozet reviewed (approved):
lgtm, thanks @cathy-zhou
